addons/binding/org.openhab.binding.innogysmarthome/src/main/java/org/openhab/binding/innogysmarthome/handler/InnogyDeviceHandler.java

```diff
@@ -0,0 +1,967 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.innogysmarthome.handler;
+
+import static org.openhab.binding.innogysmarthome.InnogyBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.OpenClosedType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.types.UpDownType;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.openhab.binding.innogysmarthome.internal.client.Constants;
+import org.openhab.binding.innogysmarthome.internal.client.entity.Property;
+import org.openhab.binding.innogysmarthome.internal.client.entity.capability.Capability;
+import org.openhab.binding.innogysmarthome.internal.client.entity.device.Device;
+import org.openhab.binding.innogysmarthome.internal.client.entity.event.Event;
+import org.openhab.binding.innogysmarthome.internal.client.entity.state.CapabilityState;
+import org.openhab.binding.innogysmarthome.internal.listener.DeviceStatusListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link InnogyDeviceHandler} is responsible for handling the {@link Device}s and their commands, which are
+ * sent to one of the channels.
+ *
+ * @author Oliver Kuhl - Initial contribution
+ */
+public class InnogyDeviceHandler extends BaseThingHandler implements DeviceStatusListener {
```

Comment:
just to clarify my earlier comment: this is just a basis for discussion, no need for you to change everything. Maybe you can give it a try and come back to me with your observations.

Reply:
I think splitting it up into a DeviceHandler for every device would lead to a lot of Handlers and I am unsure if this is really a good thing. What do you think about splitting it into an EnergyDeviceHandler, a ThermostatDeviceHandler and GeneralDeviceHandler for the rest (all based on an `abstract InnogyDeviceHandler`? This should dramatically reduce the lines of code per handler and not lead to masses of handlers.
Wdyt?